### DISCUSSION TITLE
Missing pthread dependency SofaCaribou

### DISCUSSION THREAD

**Discussion by Ziemnono** [jnbrunet/caribou#83] [Q&A] [answered]
Dear, sorry for bothering you again...
This time I am trying to compile SofaCaribou with the Caribou_With_Python3 option and it leads to the same error option in the unit tests.
```
Caribou with benchmarks
Caribou with Python support:
    Version: 3.8.10
    Executable: /usr/bin/python3.8
    Headers: /usr/include/python3.8
    Libraries: /usr/lib/x86_64-linux-gnu/libpython3.8.so
    pybind11 version: 2.6.2
    pybind11 config: /usr/lib/cmake/pybind11/pybind11Config.cmake
Caribou with SOFA support
    SOFA version: 21.12.99
    SOFA location: /home/arnaud/dev/sofa
Caribou with SofaPython3 support
    SP3 version: 20.12.0
    SP3 location: /home/arnaud/dev/sofa/plugins/SofaPython3
Caribou with OpenMP support
	Libraries: /usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so;/usr/lib/x86_64-linux-gnu/libpthread.so
Caribou.Python.Config: /home/arnaud/dev/caribou/build/lib/python3/site-packages/Caribou
Caribou.Python.Geometry: /home/arnaud/dev/caribou/build/lib/python3/site-packages/Caribou/Geometry
Caribou.Python.Topology: /home/arnaud/dev/caribou/build/lib/python3/site-packages/Caribou/Topology
SofaCaribou.Python: /home/arnaud/dev/caribou/build/lib/python3/site-packages/SofaCaribou
Configuring done
CMake Error at /snap/cmake/955/share/cmake-3.21/Modules/FindPython/Support.cmake:3316 (add_library):
  Target "SofaCaribou.Python" links to target "Threads::Threads" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  /snap/cmake/955/share/cmake-3.21/Modules/FindPython.cmake:575 (__Python_add_library)
  cmake/caribouTools.cmake:115 (python_add_library)
  src/SofaCaribou/Python/CMakeLists.txt:53 (caribou_add_python_module)


Generating done
```
I tried to add a find_package(Thread REQUIRED) in /home/arnaud/dev/caribou/src/SofaCaribou but I still obtain the same error.
Thank you for your help !

**Comment by jnbrunet** [comment]
Hey @Ziemnono ,

Could you try adding this `find_package(Thread)` in ` caribou/src/SofaCaribou/Python/CMakeLists.txt ` (before the line 53, `caribou_add_python_module(...) `) instead?

**Reply by Ziemnono** [reply]
HI @jnbrunet,
Thank you for your answer. I did it and Caribou is now expecting a path to Thread:
```
CMake Warning at src/SofaCaribou/Python/CMakeLists.txt:53 (find_package):
  By not providing "FindThread.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Thread", but
  CMake did not find one.

  Could not find a package configuration file provided by "Thread" with any
  of the following names:

    ThreadConfig.cmake
    thread-config.cmake

  Add the installation prefix of "Thread" to CMAKE_PREFIX_PATH or set
  "Thread_DIR" to a directory containing one of the above files.  If "Thread"
  provides a separate development package or SDK, be sure it has been
  installed
```
leading to the error:
```
CMake Error at /snap/cmake/955/share/cmake-3.21/Modules/FindPython/Support.cmake:3316 (add_library):
  Target "SofaCaribou.Python" links to target "Threads::Threads" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  /snap/cmake/955/share/cmake-3.21/Modules/FindPython.cmake:575 (__Python_add_library)
  cmake/caribouTools.cmake:115 (python_add_library)
  src/SofaCaribou/Python/CMakeLists.txt:54 (caribou_add_python_module)
```
I cannot locate the file ThreadConfig.cmake or FindThread.cmake, do you have any idea where it can be?

**Reply by jnbrunet** [reply]
Hey Arnaud,

I have some difficulties understanding why in your case cmake cannot find pthread, while the ubuntu 20.04 (same version as yours right?) [caribou docker image](https://github.com/jnbrunet/caribou/blob/master/ubuntu_builder_20.04.dockerfile) is able to do so. 

@guparan we could use a second pair of eyes here on this cmake matter. Do you see something that Arnaud could be missing?

**Reply by Ziemnono** [reply]
Hi @jnbrunet,
I am indeed using Ubuntu 20.04 and followed the instructions of https://caribou.readthedocs.io/en/latest/Building.html to install the prerequisites. I don't understand this mistake either...
Thank you for your time and have a nice day.

**Reply by jnbrunet** [reply]
Hey @Ziemnono , it is not a mistake on your part, I'm pretty sure. It should be quite straightforward to configure and build the project. If it is not, it means we need to improve it. I'll try to set up a docker with the same configuration as you this evening and try it on my side. I see you are using cmake from snap. Also, could you provide us your cmakecache.txt file (it should be in your Caribou's build directory). You can paste its content in a [github gist](https://gist.github.com/) and simply provide us the link afterwards.

**Reply by Ziemnono** [reply]
Hi @jnbrunet , 
Thank you for your reply and effort in trying to solve this matter quickly! Here is my gist, tell me if anything is more for accessing it.
https://gist.github.com/Ziemnono/c1966b8f7343d2bcba513123a028c284
Have a nice day.

**Reply by jnbrunet** [reply]
Hey @Ziemnono ,

Can you try with
```cmake
find_package(Threads REQUIRED)
```
instead of 
```cmake
find_package(Thread REQUIRED)
```
(notice the `s` at the end of thread?)

I'm very sorry for this typo !

**Comment by Ziemnono** [comment]
Hi @jnbrunet, I didn't even notice the typo!
After making the change I still obtain this error:
```
CMake Error at /snap/cmake/955/share/cmake-3.21/Modules/FindPython/Support.cmake:3316 (add_library):
  Target "SofaCaribou.Python" links to target "Threads::Threads" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
Call Stack (most recent call first):
  /snap/cmake/955/share/cmake-3.21/Modules/FindPython.cmake:575 (__Python_add_library)
  cmake/caribouTools.cmake:115 (python_add_library)
  src/SofaCaribou/Python/CMakeLists.txt:54 (caribou_add_python_module)
```

Still this target problem mentioned in: https://github.com/jnbrunet/caribou/discussions/78.
Thank you and have a nice day.

**Comment by guparan** [comment, marked as answer]
Hi @Ziemnono 

### Quick fix
Put `find_package(Threads REQUIRED)` early in Caribou's main CMakeLists.
For example here: https://github.com/jnbrunet/caribou/blob/master/CMakeLists.txt#L16

### Real fix
This issue is due to the removal of boost::thread done in SOFA in https://github.com/sofa-framework/sofa/pull/2264
It has been seen before and quick-fixed for one plugin: https://github.com/sofa-framework/sofa/pull/2302
The real fix has yet to be done in SOFA. It would be to disable this Threads::Threads dependency (due to GTest?) or to simply do the `find_package(Threads REQUIRED)` somewhere in the *Config.cmake.in files.

Thanks for your patience.

**Comment by Ziemnono** [comment]
Hi @guparan, thank you for this quick fix it did the job :)
I am still struggling with a scene but I will create another discussion for this thank you very much !
